### DISCUSSION TITLE
ESP32-S3 firmware

### DISCUSSION THREAD

**Discussion by smhodge42** [micropython/micropython#10206] [ESP32] [answered]
I don't have a problem erasing flash and loading firmware on an ESP32-S3 on a Espressif ESP32-S3-DevKitC-1 board, using the instructions on the Micropython website. No errors are reported.  But when I try to open a REPL session I get a continuous, non-stop display of "invalid header: 0xffffffff" messages. I used the firmware file "GENERIC_S3-20220618-v1.19.1.uf2", and also tried the ".bin" version. I tried decreasing the upload baud rate to 115,200. There no longer seems to be a "-- verify" option in <esptool> but I assume it now automatically verifies because I get a "hash of data verified" message. <esptool> version is 4.4 and the command lines used were:

esptool.py --port com14 erase_flash
esptool.py --port com14 --baud 115200 --chip esp32-s3 write_flash -z 0x1000 <firmware file name>

The development board is a genuine Espressif board, not a knock-off, purchased a couple of months ago and it worked just fine out of the box, using Visual Studio Code and Platform I/O to load a small "blink led" C++ program. 

But I really want to use Micropython, having used it on pyboards for a few years, with never a problem loading firmware on them. Any suggestions will be much appreciated.

**Comment by davefes** [comment]
Try:
  ```py
esptool.py --chip esp32 --port com14 erase_flash
```
and then
```py
esptool.py --chip esp32 --port com14 --baud 460800 write_flash -z 0x1000 esp32-20220618-v1.19.1.bin
```
change to your .bin file.  Also, you might need to change the --chip to your variant.

**Reply by smhodge42** [reply]
It gives the error messasge:
"A fatal error occurred: This chip is ESP32-S3 not ESP32. Wrong --chip argument?"
...which is correct, since it is indeed an S3 version. If I change it to "esp32-s3" (or "esp32s3", curiously both work), it works just fine.  Except that it doesn't work when I try to open a REPL session. That's the problem. I can erase flash and load new firmware just fine.

**Reply by smhodge42** [reply]
Here is the output, looks pretty convincing to me:

esptool.py v4.4
Serial port COM14
Connecting....
Chip is ESP32-S3 (revision v0.1)
Features: WiFi, BLE
Crystal is 40MHz
MAC: f4:12:fa:e6:49:40
Uploading stub...
Running stub...
Stub running...
Configuring flash size...
Flash will be erased from 0x00001000 to 0x0015cfff...
Compressed 1421968 bytes to 940979...
Wrote 1421968 bytes (940979 compressed) at 0x00001000 in 83.2 seconds (effective 136.8 kbit/s)...
Hash of data verified.

Leaving...
Hard resetting via RTS pin...

**Reply by smhodge42** [reply]
Now I am getting a different message when I attempt to open a REPL session on COM14:

ESP-ROM:esp32s3-20210327
Build:Mar 27 2021
rst:0x7 (TG0WDT_SYS_RST),boot:0x8 (SPI_FAST_FLASH_BOOT)
Saved PC:0x40043ac8
SPIWP:0xee
mode:DIO, clock div:1
load:0x3fce3808,len:0x43c
load:0x403c9700,len:0xbec
load:0x20736177,len:0x6c6c6163
Invalid image block, can't boot.
ets_main.c 329 
ESP-ROM:esp32s3-20210327
Build:Mar 27 2021
rst:0x10 (RTCWDT_RTC_RST),boot:0x8 (SPI_FAST_FLASH_BOOT)
SPIWP:0xee
mode:DIO, clock div:1
load:0x3fce3808,len:0x43c
load:0x403c9700,len:0xbec
load:0x20736177,len:0x6c6c6163
Invalid image block, can't boot.
ets_main.c 329 
....
repeating forever

Maybe the "invalid image block, can't boot" message is a clue?

Maybe the "GENERIC_S3" firmware file (see original post) doesn't work for this DevKitC-1 board?

**Reply by Wind-stormger** [reply]
Wrong flash address.

**Comment by mattytrentini** [comment]
Explicitly which [ESP32-S3-DevKitC-1](https://docs.espressif.com/projects/esp-idf/en/latest/esp32s3/hw-reference/esp32s3/user-guide-devkitc-1.html#ordering-information) board is it? 

We've had a few reports that the Octal SPI (for flash/psram) variants have had some issues. I'd actually thought that they should work with standard (quad) SPI but it may not be the case...

**Comment by smhodge42** [comment]
It is the N8R8 version, with ESP32-S3-WROOM-1-N8R8 module.

**Comment by Wind-stormger** [comment]
I am familiar with this problem, lucky.
ESP32-S3-WROOM-1U-N8R8 ,flash: 8 MB (Quad SPI),PSRAM: 8 MB (Octal SPI)
Currently, you can only download this firmware: https://micropython.org/download/GENERIC_S3/
Choose the file with the extension `.bin` to download.
Connect the USB cable to the USB port on the development board, not the UART port.
Press and hold the BOOT button, press the RESET button once, and release the BOOT button, which makes the chip enter the bootloarder mode.
Do it by esptool:
```
esptool.py --chip esp32s3 --port COM14 --baud 460800 write_flash -z 0x0 Xxxxxxxxv1.19.1.bin
```
Pay attention to replace the correct serial port number and file name.

**Reply by Wind-stormger** [reply]
I'm pretty sure this firmware will work because it doesn't use PSRAM.

**Reply by Wind-stormger** [reply]
Soon you can download the firmware using Octal SPIRAM. https://github.com/micropython/micropython/pull/8689

**Reply by sheetaljain01** [reply]
Hi, 

I have created a custom firmware for the ESP32-S3 and obtained the following files: `micropython.bin`, `bootloader.bin`, and `partition-table.bin`.

While uploading the firmware using the following command:  

python3 -m esptool --chip esp32s3 --port COM5 write_flash 0x1000 bootloader.bin 0x8000 partition-table.bin 0x10000 micropython.bin

my ESP32-S3 keeps rebooting and shows this error continuously:  

ESP-ROM:esp32s3-20210327  
Build: Mar 27 2021  
rst:0x10 (RTCWDT_RTC_RST), boot:0x8 (SPI_FAST_FLASH_BOOT)  
SPIWP:0xee  
mode: DIO, clock div:1  
load: 0x3fce3808, len: 0x43c  
load: 0x403c9700, len: 0xbec  
load: 0x20736177, len: 0x6c6c6163  
Invalid image block, can't boot.  
ets_main.c 329  

This error keeps repeating. Can you help me troubleshoot the issue?"

**Reply by projectgus** [reply]
@sheetaljain01 The bootloader offset for ESP32-S3 should be 0x0 not 0x1000. Only the original ESP32 chip uses 0x1000.

(Tip: If you do a normal build then the build process actually prints the flashing command at the end. Or you can do `make deploy BOARD=my_board PORT=portname` to have the MicroPython build system run esptool directly.)

**Comment by smhodge42** [comment]
Thanks. That worked. I have the REPL. I am curious, however. You say this firmware doesn't use PSRAM but I'm uploading exactly the same file, except to address 0x0 instead of 0x1000. In fact, yesterday I tried uploading the "normal" way to address 0x0 and it didn't work. Is it something to do with the USB port (versus the UART one) and/or manually using the bootleader?

**Reply by smhodge42** [reply]
Well, life took a detour for a year but here I am a year later trying to continue. I noticed we are now 2 micropython firmware versions later (1.21 now instead of 1.19) and now with an OCTAL-SPIRAM alternate set, for the same target as a year ago, an Espressif ESP32-S3-WROOM-1-N8R8 module on an Espressif ESP32-S3-DevKitC-1 board. So I erased the chip and installed this latest OCTAL-SPIRAM variant using:

esptool.py --chip esp32s3 --port com22 write_flash -z 0 ESP32_GENERIC_S3-SPIRAM_OCT-20231005-v1.21.0.bin

This appeared to flash ok (using the UART port) and I can open a REPL session with PuTTY and successfully use the REPL with the I2C bus so it seems like things are ok, which is great.  Except that the REPL session opened with the following error messages (in red font):

E (277) quad_psram: PSRAM ID read error: 0x00ffffff, PSRAM chip not found or not supported, or wrong PSRAM line mode
E (277) esp_psram: PSRAM enabled but initialization failed. Bailing out.

Are these anything to be concerned about? The board is supposed to have "8 MB OT PSRAM", so why these error messages?

Thanks, Steve

**Reply by smhodge42** [reply]
OOPS, I made a mistake on the above post. It was the NON octal-spiram version that gave the above (good) results:

esptool.py --chip esp32s3 --port com22 write_flash -z 0 ESP32_GENERIC_S3-20231005-v1.21.0.bin

The OCTAL-SPIRAM version (previous post) resulted in repeated rebooting (forever) when I opened a REPL session. So my additional question is why did this OCTAL-SPIRAM version NOT work at all?

**Reply by glenn20** [reply]
ESP32S3 devices should have their firmware written to offset 0 (-z 0x0) (see the flashing instructions at https://micropython.org/download/ESP32_GENERIC_S3/). The original ESP32 devices required firmware to be written at offset 0x1000. This is very likely the source of your "invalid header" errors, as the device is expecting a bootloader at offset 0x0.

**Reply by smhodge42** [reply]
I did write to address 0, see above.

**Reply by sophiedegran** [reply]
On Esp32s3 dev-kit-c1 N8R8 the 8MB flash are QD and 8MB psram are OT
On esp32s3 dev-kit-c1 N16R8 and N32R8 flash is OT (and psram OT)
As they have been changed between 1.21 and 1.22 on making image I suggest you to try a dayly image (I think it should works)

**Reply by smhodge42** [reply, marked as answer]
I got the same results with the latest daily versions.

However, **the problem is solved**. I had accidentally used PSRAM pins 36 and 37 for an I2C bus and when originally setting that up I had missed the (obvious) note in Espressif's documentation for the DevKitC about not using those pins when the module has PSRAM. My bad. I have ordered an N8 (no PSRAM) version of the DevKitC which should allow me to use my existing setup until I get around to moving those I2C lines somewhere else.

Thanks for everyone's help.

**Comment by smhodge42** [comment]
I finally got around to trying the N8 (no PSRAM) version, but I still get the same error messages (in red) when the board is reset:
    E (287) quad_psram: PSRAM ID read error: 0x00ffffff, PSRAM chip not found or not supported, or wrong PSRAM line mode
    E (288) esp_psram: PSRAM enabled but initialization failed. Bailing out.
I am using the latest firmware v1.22.1 (non Octal on the N8), on an Espressif (not a clone) ESP32-S3 DevKitC-1, not plugged into any motherboard.
It seems like Micropython is unable to gracefully know when PSRAM is (legitimately) not present.
Is that correct?

**Reply by Ruisute** [reply]
Hi, sorry for bothering; did you make it work for micropython? im doing a proyect about build a "beginners Car" (MIR in other words) for university, i tried to use VSC enviroment (Pymakr) but it didnt work, i have the same logs in terminal and it says that "safe boot cant be done",

i have the same esp32s3 kit (the one with 2 usb C ports).

**Comment by Ruisute** [comment]
Well guys, i could finally flash my esp32s3, just change the 0x1000 by 0

**Comment by guavuavcycayfayqvqu** [comment]
Same problem with esp32 s3 china clone. (Note  : flash at 0x00 not 0x10000)
